### PR TITLE
compiler: support clear

### LINKED ISSUE #3091: Support clear() in the compiler
With Go 1.21 we can add `clear()` support to the compiler that'd be especially useful for maps (`CLEARITEMS` does the job). Slices can be supported too, but some additional code would need to be generated for them.

### PR DESCRIPTION
**Pull request by Turalchik**
Close #3091.

### CODE REVIEW

File: docs/compiler.md [hunk 1]
```diff
@@ -37,7 +37,6 @@ a dialect of Go rather than a complete port of the language:
  * `comparable` is not supported
  * arrays (`[4]byte`) are not supported (https://github.com/nspcc-dev/neo-go/issues/3524)
  * `min()` and `max()` are supported for integer types only.
- * `clear()` is not supported (https://github.com/nspcc-dev/neo-go/issues/3091)
```
Comment (AnnaShaleva) [inline docs/compiler.md:40]: We still need a remark about compatibility. It doesn't work with nested arrays the same way as Go's built-in.

File: pkg/compiler/builtin_test.go [hunk 2]
```diff
@@ -105,3 +105,112 @@ func TestMaxMin(t *testing.T) {
...
+		{
+			name: "invalid type array",
+			src: `
+				package foo
+				func Main() [42]int {
+					var m = [42]int{}
+					clear(m)
+					return m
+				}
+			`,
```
Comment (AnnaShaleva) [inline pkg/compiler/builtin_test.go]: Here and below:
```suggestion
				}`,
```

File: pkg/compiler/builtin_test.go [hunk 3]
```diff
@@ -105,3 +105,112 @@ func TestMaxMin(t *testing.T) {
...
+	tests := []struct {
+		name        string
+		src         string
+		expectedErr string
+		expected    any
+	}{
+		{
+			name: "invalid type array",
+			src: `
+				package foo
```
Comment (AnnaShaleva) [inline pkg/compiler/builtin_test.go]: Here and below:
```suggestion
			src: `package foo
```

File: pkg/compiler/builtin_test.go [hunk 4]
```diff
@@ -105,3 +105,112 @@ func TestMaxMin(t *testing.T) {
...
+				func Main() [42]int {
+					var m = [42]int{}
```
Comment (AnnaShaleva) [inline pkg/compiler/builtin_test.go:120]: Just for the record: arrays are not supported yet, ref. https://github.com/nspcc-dev/neo-go/issues/3524.

But this test should be kept, it's a valid one.

File: pkg/compiler/builtin_test.go [hunk 5]
```diff
@@ -105,3 +105,112 @@ func TestMaxMin(t *testing.T) {
...
+				func Main() []string {
+					var s = []string{"one", "two", "three"}
+					clear(s)
+					return s
+				}
+			`,
+			expected: []stackitem.Item{stackitem.Make(""), stackitem.Make(""), stackitem.Make("")},
+		},
+		{
+			name: "slice of structs",
```
Comment (AnnaShaleva) [inline pkg/compiler/builtin_test.go:161]: Add "slice of maps" test-case.

File: pkg/compiler/codegen.go [hunk 6]
```diff
@@ -2010,6 +2010,32 @@ func (c *codegen) convertBuiltin(expr *ast.CallExpr) {
 		for range len(expr.Args) - 1 {
 			emit.Opcodes(c.prog.BinWriter, opcode.MAX)
 		}
+	case "clear":
+		switch typ := c.typeOf(expr.Args[0]).(type) {
+		case *types.Map:
+			emit.Opcodes(c.prog.BinWriter, opcode.CLEARITEMS)
+		case *types.Slice:
+			emit.Opcodes(c.prog.BinWriter, opcode.DUP, opcode.SIZE, opcode.DEC,
```
Comment (AnnaShaleva) [inline pkg/compiler/codegen.go]: Need a test-case for empty slice, I need to ensure that it works properly with `-1` returned by `DEC`.

Comment (AnnaShaleva) [reply pkg/compiler/codegen.go]: Also for an empty map.

File: docs/compiler.md (hunk 1 above)
Comment (Turalchik) [reply docs/compiler.md:40]: I didn't quite understand how our implementation of `clear()` differs from the built-in Go? I added a test for nested slices.

Comment (AnnaShaleva) [reply docs/compiler.md:40]: > If the argument type is a type parameter, the type parameter's type set must contain only map or slice types, and clear performs the operation implied by the type argument.

I was confused by this line in the documentation and thought that clear is invoked recursively for elements of nested slices. But it's not true, you're right, elements of nested slice are set to zero value which is `nil` in case of a slice.

So this line should just be removed.

File: pkg/compiler/builtin_test.go [hunk 7]
```diff
@@ -105,3 +105,149 @@ func TestMaxMin(t *testing.T) {
...
+			src: `package foo
+				func Main() []any {
+					var s = []any{}
+					clear(s)
+					return s
+				}`,
+			expected: []stackitem.Item{},
+		},
+		{
+			name: "empty slice",
```
Comment (AnnaShaleva) [inline pkg/compiler/builtin_test.go]: It's `empty map`.

File: pkg/compiler/codegen.go [hunk 8]
```diff
@@ -2010,6 +2010,32 @@ func (c *codegen) convertBuiltin(expr *ast.CallExpr) {
...
+			emit.Opcodes(c.prog.BinWriter, opcode.PUSH0)
+			clearItemsEnd := c.newLabel()
+			emit.Jmp(c.prog.BinWriter, opcode.JMPLTL, clearItemsEnd)
+			emit.Opcodes(c.prog.BinWriter, opcode.PUSH2, opcode.PICK, opcode.PUSH2,
+				opcode.PICK, opcode.PUSH2, opcode.PICK, opcode.SETITEM,
+				opcode.SWAP, opcode.DEC, opcode.SWAP, opcode.OVER,
+			)
+			emit.Jmp(c.prog.BinWriter, opcode.JMPL, clearItemsBegin)
+			c.setLabel(clearItemsEnd)
+			emit.Opcodes(c.prog.BinWriter, opcode.DROP, opcode.DROP, opcode.DROP)
```
Comment (AnnaShaleva) [inline pkg/compiler/codegen.go]: This program looks good to me, I doubt we can improve it in terms of cost or the number of the emitted instructions. But I'd ask @roman-khimov's opinion before the merge, just in case.

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/compiler/builtin_test.go [hunk 9]
```diff
@@ -105,3 +105,149 @@ func TestMaxMin(t *testing.T) {
...
+				
+				func Main() []S1 {
+					var s = []S1 {
+						{
+							S: S2 {
+								A: 1,
+								B: "str",
+								C: true,
+							},
+						},
```
Comment (roman-khimov) [inline pkg/compiler/builtin_test.go:188]: Add another element here, then `clear(s)`, then `s[0].S.A = 42`, then check for `s[1].S.A`.

File: pkg/compiler/codegen.go (hunk 8 above)
Comment (roman-khimov) [reply pkg/compiler/codegen.go]: CLEARITEMS + APPEND in the loop can cost less for large arrays (CLEARITEMS itself costs something, but then it's one element less to manage in the loop).

Comment (AnnaShaleva) [reply pkg/compiler/codegen.go]: Let's then move to this approach, it's easier to maintain.

File: pkg/compiler/builtin_test.go (hunk 9 above)
Comment (roman-khimov) [reply pkg/compiler/builtin_test.go:188]: Have you tried exactly the scenario outlined above (can be a separate test case)? I think there is going to be a problem.

Comment (AnnaShaleva) [reply pkg/compiler/builtin_test.go:188]: @roman-khimov is right, structure is a reference type, so we can't use `DUP` over default element; instead we need to construct default element every time we need it.

Comment (roman-khimov) [reply pkg/compiler/builtin_test.go:188]: And it'd be cheaper for basic types as well, PUSH0/PUSHNULL and alike are very cheap.

**Review by roman-khimov — APPROVED** [review]
And the bytecode is much simpler now.
